### PR TITLE
Refactor listeners and authentication configuration

### PR DESCRIPTION
**Pull request by juan131**
Signed-off-by: juan131 <juanariza@vmware.com>

**Description of the change**

This PR refactors how listeners and authentication is configured on Kafka, allowing users to have different listeners for different purposes, and configure different authentication protocols on them.

The supported listeners will be:

* INTERNAL: used for inter-broker communications.
* CLIENT: used for communications with clients that are within the same network as Kafka brokers.
* EXTERNAL: used for communications with clients that are on a different network.

To use a more complex configuration, users will have to mount their own `server.properties` and `kafka_jaas.conf` configuration files.

**Benefits**

We can achieve more complex configurations without assuming so many configuration details.

**Possible drawbacks**

Users will have to adapt their Docker Compose and scripts files to the new format.

**Applicable issues**

None

**Additional information**

A proposal to adapt the Kafka Helm chart according to these changes can be found here: https://github.com/bitnami/charts/pull/2708

### CODE REVIEW

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 1]
```diff
@@ -304,66 +356,75 @@ kafka_configure_ssl_listener() {
...
+    info "Configuring Kafka for client communications with $protocol authentication."
+
+    if [[ " ${allowed_protocols[@]} " =~ " $protocol " ]]; then
+        kafka_server_conf_set security.inter.broker.protocol $protocol
+        if [[ "$protocol" = "PLAINTEXT" ]]; then
+            warn "Client communications are configured as PLAINTEXT. This is not safe for production environments."
+        fi
+        if [[ "$protocol" = "SASL_PLAINTEXT" ]] || [[ "$protocol" = "SASL_SSL" ]]; then
+            # TODO (juan131): Support other SASL implementations (e.g. GSSAPI)
+            # Please do not confuse SASL/PLAIN with PLAINTEXT (see https://docs.confluent.io/current/kafka/authentication_sasl/authentication_sasl_plain.html#sasl-plain-overview)
```
Comment (marcosbc) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:415]: Same as above.

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 2]
```diff
@@ -304,66 +356,75 @@ kafka_configure_ssl_listener() {
...
-    kafka_producer_consumer_conf_set sasl.mechanism PLAIN
+kafka_configure_client_communications() {
+    local -r protocol="${1:?missing environment variable protocol}"
+    local -r allowed_protocols=("PLAINTEXT" "SASL_PLAINTEXT" "SASL_SSL" "SSL")
+    info "Configuring Kafka for client communications with $protocol authentication."
+
+    if [[ " ${allowed_protocols[@]} " =~ " $protocol " ]]; then
+        kafka_server_conf_set security.inter.broker.protocol $protocol
+        if [[ "$protocol" = "PLAINTEXT" ]]; then
+            warn "Client communications are configured as PLAINTEXT. This is not safe for production environments."
```
Comment (marcosbc) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:411]:
```suggestion
            warn "Client communications are configured using PLAINTEXT listeners. For safety reasons, do not use this in a production environment."
```

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 3]
```diff
@@ -304,66 +356,75 @@ kafka_configure_ssl_listener() {
...
+    info "Configuring Kafka for inter-broker communications with $protocol authentication."
+
+    if [[ " ${allowed_protocols[@]} " =~ " $protocol " ]]; then
+        kafka_server_conf_set security.inter.broker.protocol $protocol
+        if [[ "$protocol" = "PLAINTEXT" ]]; then
+            warn "Inter-broker communications are configured as PLAINTEXT. This is not safe for production environments."
+        fi
+        if [[ "$protocol" = "SASL_PLAINTEXT" ]] || [[ "$protocol" = "SASL_SSL" ]]; then
+            # TODO (juan131): Support other SASL implementations (e.g. GSSAPI)
+            # Please do not confuse SASL/PLAIN with PLAINTEXT (see https://docs.confluent.io/current/kafka/authentication_sasl/authentication_sasl_plain.html#sasl-plain-overview)
```
Comment (marcosbc) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:379]: I would remove the TODO and adding some wording improvements, i.e.:

```suggestion
            # The below lines would need to be updated to support other SASL implementations (i.e. GSSAPI)
            # IMPORTANT: Do not confuse SASL/PLAIN with PLAINTEXT
            # For more information, see: https://docs.confluent.io/current/kafka/authentication_sasl/authentication_sasl_plain.html#sasl-plain-overview)
```

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 4]
```diff
@@ -304,66 +356,75 @@ kafka_configure_ssl_listener() {
...
-    kafka_server_conf_set security.inter.broker.protocol SSL
-    # Set producer/consumer configuration
-    kafka_producer_consumer_conf_set security.protocol SSL
+kafka_configure_internal_communications() {
+    local -r protocol="${1:?missing environment variable protocol}"
+    local -r allowed_protocols=("PLAINTEXT" "SASL_PLAINTEXT" "SASL_SSL" "SSL")
+    info "Configuring Kafka for inter-broker communications with $protocol authentication."
+
+    if [[ " ${allowed_protocols[@]} " =~ " $protocol " ]]; then
+        kafka_server_conf_set security.inter.broker.protocol $protocol
```
Comment (marcosbc) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:373]:
```suggestion
    info "Configuring Kafka for inter-broker communications with ${protocol} authentication."

    if [[ " ${allowed_protocols[@]} " =~ " ${protocol} " ]]; then
        kafka_server_conf_set security.inter.broker.protocol "$protocol"
```

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 5]
```diff
@@ -304,66 +356,75 @@ kafka_configure_ssl_listener() {
...
+            kafka_server_conf_set sasl.mechanism.inter.broker.protocol PLAIN
+        fi
+        if [[ "$protocol" = "SASL_SSL" ]] || [[ "$protocol" = "SSL" ]]; then
+            kafka_configure_ssl
+            # We need to enable 2 way authentication on SASL_SSL so brokers authenticate each other.
+            # It won't affect client communications unless the SSL protocol is for them.
+            kafka_server_conf_set ssl.client.auth required
+        fi
+    else
+        error "Authentication protocol $protocol is not supported!"
```
Comment (marcosbc) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:389]:
```suggestion
        error "Authentication protocol ${protocol} is not supported!"
```

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 6]
```diff
@@ -304,66 +356,75 @@ kafka_configure_ssl_listener() {
...
+            kafka_server_conf_set sasl.mechanism.inter.broker.protocol PLAIN
+        fi
+        if [[ "$protocol" = "SASL_SSL" ]] || [[ "$protocol" = "SSL" ]]; then
+            kafka_configure_ssl
+        fi
+        if [[ "$protocol" = "SSL" ]]; then
+            kafka_server_conf_set ssl.client.auth required
+        fi
+    else
+        error "Authentication protocol $protocol is not supported!"
```
Comment (marcosbc) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:425]:
```suggestion
        error "Authentication protocol ${protocol} is not supported!"
```

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 7]
```diff
@@ -201,84 +199,139 @@ kafka_create_alias_environment_variables() {
...
 #########################
 kafka_generate_jaas_authentication_file() {
+    local -r internal_protocol="${1:?missing environment variable internal_protocol}"
+    local -r client_protocol="${2:?missing environment variable client_protocol}"
+
     if [[ ! -f "${KAFKA_CONF_DIR}/kafka_jaas.conf" ]]; then
         info "Generating JAAS authentication file"
-        render-template > "${KAFKA_CONF_DIR}/kafka_jaas.conf" <<EOF
+        if [[ "$client_protocol" =~ "SASL" ]]; then
+            render-template >> "${KAFKA_CONF_DIR}/kafka_jaas.conf" <<EOF
```
Comment (marcosbc) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:277]: If I'm not wrong, `render-template` requires all the environment variables to exist, or it will throw an error. Maybe we could do this instead? (and avoid having to rely on an external tool):

```console
$ cat >> "${KAFKA_CONF_DIR}/kafka_jaas.conf" <<EOF

KafkaClient {
   org.apache.kafka.common.security.plain.PlainLoginModule required
   org.apache.kafka.common.security.plain.PlainLoginModule required
   username="${KAFKA_CLIENT_USER:-}"
   password="${KAFKA_CLIENT_PASSWORD:-}";
   };
EOF
```

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 8]
```diff
@@ -201,84 +199,139 @@ kafka_create_alias_environment_variables() {
...
 
    org.apache.kafka.common.security.scram.ScramLoginModule required;
-};
+   };
 EOF
-        if [[ -n "$KAFKA_ZOOKEEPER_USER" ]] && [[ -n "$KAFKA_ZOOKEEPER_PASSWORD" ]]; then
-            info "Configuring ZooKeeper client credentials"
+        elif [[ "$client_protocol" =~ "SASL" ]]; then
             render-template >> "${KAFKA_CONF_DIR}/kafka_jaas.conf" <<EOF
+KafkaServer {
```
[the remark this answers is not in the document]
Comment (juan131) [reply 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:299]: Correct, just following the same style used on Kafka examples

File: README.md [hunk 9]
```diff
@@ -7,7 +6,7 @@ thousands of companies. Kafka requires a connection to a Zookeeper service.
 
 [https://kafka.apache.org/](https://kafka.apache.org/)
 
-# TL;DR;
+## TL;DR;
```
[the remark this answers is not in the document]
Comment (juan131) [reply README.md:9]: I think we should only have one "H1" heading in the whole README.md to meet best practices for markdown

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 10]
```diff
@@ -201,84 +201,140 @@ kafka_create_alias_environment_variables() {
...
+
+    if [[ ${KAFKA_CFG_LISTENERS:-} =~ INTERNAL:\/\/:([0-9]*) ]]; then
+        internal_port="${BASH_REMATCH[1]}"
+        check_allowed_listener_port "$internal_port"
+    fi
+    if [[ ${KAFKA_CFG_LISTENERS:-} =~ CLIENT:\/\/:([0-9]*) ]]; then
+        client_port="${BASH_REMATCH[1]}"
+        check_allowed_listener_port "$client_port"
+    fi
+    [[ -n ${internal_port:-} && -n ${client_port:-} ]] && check_conflicting_listener_ports "$internal_port" "$client_port"
```
Comment (marcosbc) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:238]: For consistency, we should use quotes to contain environment variables:

```suggestion
    if [[ "${KAFKA_CFG_LISTENERS:-}" =~ INTERNAL:\/\/:([0-9]*) ]]; then
        internal_port="${BASH_REMATCH[1]}"
        check_allowed_listener_port "$internal_port"
    fi
    if [[ "${KAFKA_CFG_LISTENERS:-}" =~ CLIENT:\/\/:([0-9]*) ]]; then
        client_port="${BASH_REMATCH[1]}"
        check_allowed_listener_port "$client_port"
    fi
    [[ -n "${internal_port:-}" && -n "${client_port:-}" ]] && check_conflicting_listener_ports "$internal_port" "$client_port"
```

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 11]
```diff
@@ -304,66 +359,77 @@ kafka_configure_ssl_listener() {
...
-    # Set Kafka configuration
-    kafka_server_conf_set security.inter.broker.protocol SSL
-    # Set producer/consumer configuration
-    kafka_producer_consumer_conf_set security.protocol SSL
+kafka_configure_internal_communications() {
+    local -r protocol="${1:?missing environment variable protocol}"
+    local -r allowed_protocols=("PLAINTEXT" "SASL_PLAINTEXT" "SASL_SSL" "SSL")
+    info "Configuring Kafka for inter-broker communications with ${protocol} authentication."
+
+    if [[ " ${allowed_protocols[@]} " =~ " ${protocol} " ]]; then
```
Comment (marcosbc) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:375]: I think this line could give some shellcheck issues. I think the bellow is a better approach:

```suggestion
    if [[ " ${allowed_protocols[*]} " =~ " ${protocol} " ]]; then
```

File: 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh [hunk 12]
```diff
@@ -304,66 +359,77 @@ kafka_configure_ssl_listener() {
...
-    kafka_server_conf_set security.inter.broker.protocol SASL_PLAINTEXT
-    # Set producer/consumer configuration
-    kafka_producer_consumer_conf_set security.protocol SASL_PLAINTEXT
-    kafka_producer_consumer_conf_set sasl.mechanism PLAIN
+kafka_configure_client_communications() {
+    local -r protocol="${1:?missing environment variable protocol}"
+    local -r allowed_protocols=("PLAINTEXT" "SASL_PLAINTEXT" "SASL_SSL" "SSL")
+    info "Configuring Kafka for client communications with ${protocol} authentication."
+
+    if [[ " ${allowed_protocols[@]} " =~ " ${protocol} " ]]; then
```
Comment (marcosbc) [inline 2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh:412]: Same

```suggestion
    if [[ " ${allowed_protocols[*]} " =~ " ${protocol} " ]]; then
```

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (marcosbc) [reply README.md:9]: In any case, that's a change I think we should consider making for all images at once, so it is consistent.

However, my main concern with this is that the "main" title is titled "What is Kafka?". I.e.:

* What is Kafka?

  * TL;DR;
  * Why use Bitnami Images?

It just looks strange. IMHO we should change the main section to "Bitnami Docker Kafka" or something similar, and then add "What is Kafka" as H2.
